task-maker-async/src/store.rs

```diff
@@ -0,0 +1,114 @@
+pub struct InputFileWriteHandle(usize);
+#[derive(Debug, Serialize, Deserialize)]
+pub struct ComputationWriteHandle(usize);
+#[derive(Debug, Serialize, Deserialize)]
+pub struct ReadHandle(usize);
+
+#[derive(Debug, Serialize, Deserialize)]
+pub enum ComputationDataFile {
+    Outcome,
+    OutputFile(String),
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub enum FileReadingOutcome {
+    /// The file has been deleted, for example because the worker responsible for the execution has
+    /// disappeared.
+    Dropped,
+    /// The file has been fully read.
+    EndOfFile,
+    /// A new chunk of data is available.
+    Data(Vec<u8>),
+}
+
+#[tarpc::service]
+pub trait Store {
+    /// Used by the client to upload inputs to the DAG. Returns None if the file is already
+    /// present. Either way, this leases the corresponding hash for LEASE_LENGTH_SEC.
+    async fn create_input_file(hash: InputFileHash) -> Option<InputFileWriteHandle>;
+
+    /// Appends a chunk to the input file; the InputFileWriteHandle must have been obtained through
+    /// create_input_file. Renews the lease on the corresponding hash.
+    async fn append_input_chunk(file_handle: InputFileWriteHandle, data: Vec<u8>)
+        -> Result<(), ()>;
+
+    /// Finalizes the input file; no more data can be appended, and the resulting hash must match
+    /// the claimed hash when create_input_file was called.
+    async fn finalize_input_write_file(file_handle: InputFileWriteHandle) -> Result<(), ()>;
+
+    /// Refreshes the lease for the given input file, or creates one if none was present.
+    /// It is guaranteed that the input file will not be deleted while there's an outstanding lease
+    /// to it. It is an error to refresh a lease of a non-existent input file.
+    async fn refresh_input_lease(hash: InputFileHash) -> Result<(), ()>;
+
+    /// Refreshes the given computation writing lease. Used by workers while they're running an
+    /// evaluation group.
+    async fn refresh_computation_write_lease(handle: ComputationWriteHandle) -> Result<(), ()>;
+
+    /// Appends data to a file in the output of a computation. The file is created if it does not
+    /// exist. Refreshes the writing lease.
+    async fn append_computation_chunk(
+        handle: ComputationWriteHandle,
+        file: ComputationDataFile,
+    ) -> Result<(), ()>;
+
+    /// Finalizes a computation output. Terminates the writing lease.
+    async fn finalize_computation(handle: ComputationWriteHandle);
+
+    /// Opens an input file for reading. Creates a lease for the input file that will prevent it
+    /// from being dropped. If the file is not present, returns an error immediately.
+    async fn read_input_file(hash: InputFileHash) -> Result<ReadHandle, ()>;
+
+    /// Opens a computed file for reading. Creates a lease for the computation data that will prevent it
+    /// from being dropped. If the file is not present, waits until the file is created.
```

Comment:
What if the file is not finalized? If we can open nonfinalized files, then what if the file is never finalized (e.g. because its writer died)?

Reply:
Then attempts at reading it will return FileReadingOutcome::Dropped.  Makes sense?

Reply:
So, we'll read until the worker dies, and then we'll fail the next read, regardless of where in the file we were trying to read?

Reply:
Yep -- see also other answer (note that until a computation is finalized, there is no EOF)